chef/cookbooks/nova/recipes/ceph.rb

```diff
@@ -0,0 +1,162 @@
+      next
+    end
+
+    cmd = ["ceph", "-k", admin_keyring, "-c", ceph_conf, "-s"]
+    check_ceph = Mixlib::ShellOut.new(cmd)
+
+    unless check_ceph.run_command.stdout.match("(HEALTH_OK|HEALTH_WARN)")
+      Chef::Log.info("Ceph cluster is not healthy; skipping the ceph setup for backend #{volume[:backend_name]}")
+      next
+    end
+  end
+
+  rbd_user = volume[:rbd][:user]
+  rbd_uuid = volume[:rbd][:secret_uuid]
+
+  secret_file_path = "/etc/ceph/ceph-secret-#{volid}.xml"
+  
+  file secret_file_path do
+    owner "root"
+    group "root"
+    mode "0640"
+    content "<secret ephemeral='no' private='no'> <uuid>#{rbd_uuid}</uuid><usage type='ceph'> <name>client.#{rbd_user} secret</name> </usage> </secret>"
+  end #file secret_file_path
+
+  ruby_block "save nova key as libvirt secret" do
+    block do
+      if system("virsh hostname &> /dev/null")
+        # First remove conflicting secrets due to same usage name
+        secret_list = %x[ virsh secret-list 2> /dev/null ]
+
+        secret_lines = secret_list.split("\n")
+        if secret_lines.length < 2 || !secret_lines[0].start_with?("UUID") || !secret_lines[1].start_with?("----")
+          raise "cannot fetch list of libvirt secret"
+        end
+        secret_lines.shift(2)
+
+        secret_lines.each do |secret_line|
+          secret_uuid = secret_line.split(" ")[0]
+          secret_xml = %x[ virsh secret-dumpxml #{secret_uuid} ]
+          # some secrets might not be ceph-related, skip these
+          next if secret_xml.index("<usage type='ceph'>").nil?
+
+          # lazy xml parsing
+          re_match = %r[<usage type='ceph'>.*<name>(.*)</name>]m.match(secret_xml)
+          next if re_match.nil?
+          secret_usage = re_match[1]
+
+          undefine = false
+
+          if secret_uuid == rbd_uuid
+            undefine = true if secret_usage != "client.#{rbd_user} secret"
+          else
+            undefine = true if secret_usage == "client.#{rbd_user} secret"
+          end
+
+          if undefine
+            %x[ virsh secret-undefine #{secret_uuid} ]
+          end
+        end
+
+        # Now add our secret and its value
+        client_key = %x[ ceph auth get-key client.'#{rbd_user}' ]
+        raise 'getting nova client key failed' unless $?.exitstatus == 0
+
+        secret = %x[ virsh secret-get-value #{rbd_uuid} 2> /dev/null ].chomp.strip
+        if secret != client_key
+          %x[ virsh secret-define --file '#{secret_file_path}' ]
+          raise 'generating secret file failed' unless $?.exitstatus == 0
+
+          %x[ virsh secret-set-value --secret '#{rbd_uuid}' --base64 '#{client_key}' ]
+          raise 'importing secret file failed' unless $?.exitstatus == 0
+        end
+      end
+    end
+  end
+
+  if node['nova']['rbd']['user'] != rbd_user || node['nova']['rbd']['secret_uuid'] != rbd_uuid
+    node['nova']['rbd']['user'] = rbd_user
+    node['nova']['rbd']['secret_uuid'] = rbd_uuid
+    node.save
+  end
```

Comment:
I think that part is actually wrong: iirc, nova will always favor the local secret_uuid over what cinder gives it, so we should not define it locally, I believe.